### PR TITLE
feat: ICP Internet Identity Login

### PR DESCRIPTION
**Pull request by alenmestrov**
### Implemented ICP Login with Internet Identity (II) Integration

**Overview**  
This PR introduces Internet Identity (II) login functionality for users to authenticate via ICP. The flow is as follows:
1. The client requests challenge from the server
2. Triggers signing it with II
3. Sends signature and needed metadata to node
4. The signature is verified on the node and stored appropriately if it passes validation

**Features**  
1. **ICP Wallet Integration**
   - Added support for logging in with Internet Identity (II).
   - Implemented routes and UI components for the ICP login page.

4. **Signature Handling**
   - Signature Request: After starting login process, the client requests a challenge to sign.
   - Challenge is signed using II.
   - Signature Verification: Implemented server-side logic to verify the signature and the associated data.

5. **Login Selector**
   - Added the ICP II option to the login selector.
   - Ensured the frontend handles and displays the ICP II public key for better user visibility.

**Code Details**  
- **Frontend**: Added the logic for initiating login requests, fetching the challenge, and sending the signed messages.
- **Backend**: Integrated signature verification using II-provided public keys and cryptographic functions.

**Fixes**  
- Addressed styling adjustments for handling the ICP public key length.

https://github.com/user-attachments/assets/43234628-c9bb-4190-b493-229bd54c88d8

### CODE REVIEW

File: crates/primitives/src/identity.rs [hunk 1]
```diff
@@ -150,6 +150,12 @@ pub enum WalletType {
         #[serde(rename = "walletName")]
         wallet_name: String,
     },
+    INTERNETCOMPUTER {
```
Comment (MatejVukosav) [inline crates/primitives/src/identity.rs]: INTERNET_COMPUTER

File: crates/server/src/verifywalletsignatures/internetcomputer.rs [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+///
+/// # Returns
+/// - `Ok(())`: If the signature and delegation chain are successfully verified.
+/// - `Err(ApiError)`: If any validation step fails, such as parsing errors, signature mismatches, or invalid delegation chains.
+///
+/// # Errors
+/// This function will return an `ApiError` in case of issues like invalid input, signature mismatches, or verification failure.
+pub async fn verify_internet_identity_signature(
+    challenge: &[u8],
+    signed_delegation_chain_json: &str,
```
Comment (MatejVukosav) [inline crates/server/src/verifywalletsignatures/internetcomputer.rs]: it looks like it should be serde &Value type

File: node-ui/src/api/dataSource/NodeDataSource.ts [hunk 3]
```diff
@@ -17,6 +17,7 @@ export enum Network {
   ARB = 'ARB',
   ZK = 'ZK',
   STARKNET = 'STARKNET',
+  INTERNETCOMPUTER = 'INTERNETCOMPUTER',
```
Comment (MatejVukosav) [inline node-ui/src/api/dataSource/NodeDataSource.ts]: INTERNET_COMPUTER

File: node-ui/src/api/dataSource/NodeDataSource.ts [hunk 4]
```diff
@@ -198,6 +212,20 @@ export namespace WalletType {
   }): WalletType {
     return { type: 'STARKNET', walletName } as SNWalletType;
   }
+
+  export function INTERNETCOMPUTER({
+    cannisterId = 'rdmx6-jaaaa-aaaaa-aaadq-cai',
```
Comment (MatejVukosav) [inline node-ui/src/api/dataSource/NodeDataSource.ts]: What is this value rdmx6-jaaaa-aaaaa-aaadq-cai' ?

File: node-ui/src/api/dataSource/NodeDataSource.ts [hunk 5]
```diff
@@ -175,7 +180,16 @@ interface SNWalletType extends WalletTypeBase<'STARKNET'> {
   walletName: string;
 }
 
-export type WalletType = ETHWalletType | NEARWalletType | SNWalletType;
+interface INTERNETCOMPUTERWalletType
```
Comment (MatejVukosav) [inline node-ui/src/api/dataSource/NodeDataSource.ts]: InternetComputerWalletType

File: node-ui/src/hooks/useInternetComputer.tsx [hunk 6]
```diff
@@ -0,0 +1,352 @@
...
+import { useNavigate } from 'react-router-dom';
+import { setStorageNodeAuthorized } from '../auth/storage';
+import { useServerDown } from '../context/ServerDownContext';
+import translations from '../constants/en.global.json';
+
+interface RequestNodeDataProps {
+  setErrorMessage: (msg: string) => void;
+}
+
+interface useInternetComputerReturn {
```
Comment (MatejVukosav) [inline node-ui/src/hooks/useInternetComputer.tsx]: I don't understand from the name what is this for.  "Return" ?

File: node-ui/src/hooks/useInternetComputer.tsx [hunk 7]
```diff
@@ -0,0 +1,352 @@
...
+    url: 'https://identity.ic0.app',
+    cannisterId: 'rdmx6-jaaaa-aaaaa-aaadq-cai',
+  },
+  staging: {
+    url: 'https://beta.identity.ic0.app/',
+    cannisterId: 'fgte5-ciaaa-aaaad-aaatq-cai',
+  },
+};
+
+export type NetworkId = keyof typeof possibleNetworks;
```
Comment (MatejVukosav) [inline node-ui/src/hooks/useInternetComputer.tsx]: Should this be the reusable type?

File: node-ui/src/hooks/useInternetComputer.tsx [hunk 8]
```diff
@@ -0,0 +1,352 @@
...
+export function useInternetComputer(): useInternetComputerReturn {
+  const navigate = useNavigate();
+  const [ready, setReady] = useState<boolean>(true);
+  const { showServerDownPopup } = useServerDown();
+  const [walletSignatureData, setWalletSignatureData] =
+    useState<WalletSignatureData | null>(null);
+  const [signData, setSignData] = useState<SignData | null>(null);
+
+  const [currentNetwork, setCurrentNetwork] = useState(
+    possibleNetworks['production'],
```
Comment (MatejVukosav) [inline node-ui/src/hooks/useInternetComputer.tsx]: Is this like mainnet and testnet or something different?

File: node-ui/src/hooks/useInternetComputer.tsx [hunk 9]
```diff
@@ -0,0 +1,352 @@
...
+
+  const signMessageAndLogin = useCallback(
+    async (isLogin: boolean, setErrorMessage: (msg: string) => void) => {
+      try {
+        setErrorMessage('');
+        setReady(false);
+
+        console.log('walletSignatureData:', walletSignatureData);
+
+        // console.log('currentNetwork:', currentNetwork);
```
Comment (MatejVukosav) [inline node-ui/src/hooks/useInternetComputer.tsx]: clean

File: node-ui/src/hooks/useInternetComputer.tsx [hunk 10]
```diff
@@ -0,0 +1,352 @@
...
+        console.error(`${t.errorChangingNetwork}: ${error}`);
+        setErrorMessage(`${t.errorChangingNetwork}`);
+      }
+      setReady(true);
+    },
+    [],
+  );
+
+  const signMessageAndLogin = useCallback(
+    async (isLogin: boolean, setErrorMessage: (msg: string) => void) => {
```
Comment (MatejVukosav) [inline node-ui/src/hooks/useInternetComputer.tsx]: add return type

File: node-ui/src/hooks/useInternetComputer.tsx [hunk 11]
```diff
@@ -0,0 +1,352 @@
...
+                verifyingKey: publicKey,
+              };
+              if (walletSignatureData?.payload) {
+                const internetComputerRequest: LoginRequest = {
+                  walletSignature: JSON.stringify(delegationChain),
+                  payload: walletSignatureData.payload,
+                  walletMetadata: walletMetadata,
+                };
+
+                console.log(
```
Comment (MatejVukosav) [inline node-ui/src/hooks/useInternetComputer.tsx]: do you need this?

File: node-ui/src/hooks/useInternetComputer.tsx [hunk 12]
```diff
@@ -0,0 +1,352 @@
...
+                  'internetComputerRequest:',
+                  internetComputerRequest,
+                );
+
+                const result: ResponseData<LoginResponse> = isLogin
```
Comment (MatejVukosav) [inline node-ui/src/hooks/useInternetComputer.tsx]: it is strange that both calls return the same type

File: node-ui/src/hooks/useInternetComputer.tsx [hunk 13]
```diff
@@ -0,0 +1,352 @@
...
+        setErrorMessage(t.signMessageError);
+      } finally {
+        setReady(true);
+      }
+    },
+    [walletSignatureData, currentNetwork],
+  );
+
+  // Helper function to convert Uint8Array to hex string
+  function uint8ArrayToHexString(arr: Uint8Array) {
```
Comment (MatejVukosav) [inline node-ui/src/hooks/useInternetComputer.tsx]: Looks like it should be in utils

File: node-ui/src/hooks/useInternetComputer.tsx [hunk 14]
```diff
@@ -0,0 +1,352 @@
...
+    derivationOrigin,
+    sessionPublicKey,
+  }: {
+    url: string;
+    maxTimeToLive?: bigint;
+    derivationOrigin?: string;
+    sessionPublicKey: Uint8Array;
+  }): Promise<DelegationIdentity> => {
+    try {
+      // Figure out the II URL to use
```
Comment (MatejVukosav) [inline node-ui/src/hooks/useInternetComputer.tsx]: is this a todo or is figured out?

File: node-ui/src/utils/rootkey.ts [hunk 15]
```diff
@@ -75,32 +76,43 @@ export function mapApiResponseToObjects(
...
-          return starknetObject;
-        }
-      });
+    const rootKeys: (
+      | ETHRootKey
+      | NearRootKey
+      | StarknetRootKey
+      | InternetComputerRootKey
+    )[] = didResponse?.did?.root_keys?.map((obj: ApiRootKey) => {
+      if (obj.wallet.type === Network.NEAR) {
```
Comment (MatejVukosav) [inline node-ui/src/utils/rootkey.ts]: use switch case for multiple ifs

File: crates/primitives/src/identity.rs (hunk 1 above)
Comment (alenmestrov) [reply crates/primitives/src/identity.rs]: the thing here is that then rust throws a warning Variant `INTERNET_COMPUTER` should have CamelCase name, e.g. `InternetComputer. 
If you'd like, I can change it but then I would recommend doing it for every other variant.

File: node-ui/src/hooks/useInternetComputer.tsx (hunk 8 above)
Comment (alenmestrov) [reply node-ui/src/hooks/useInternetComputer.tsx]: Yes, in ICP context production and staging are considered mainnet and testnet

File: crates/primitives/src/identity.rs (hunk 1 above)
Comment (MatejVukosav) [reply crates/primitives/src/identity.rs]: Yes, that is standard which we didn't follow 😅 you can ignore warning for now and update them separately

File: node-ui/src/api/dataSource/NodeDataSource.ts (hunk 4 above)
Comment (alenmestrov) [reply node-ui/src/api/dataSource/NodeDataSource.ts]: this is the default value for the canister ID. There are two canisters that are used, one for production (mainnet) and one for staging (testnet) similar to the NEAR network type.

File: node-ui/src/hooks/useInternetComputer.tsx (hunk 7 above)
Comment (alenmestrov) [reply node-ui/src/hooks/useInternetComputer.tsx]: No, this type (NetworkId) is only relevant in the context of the Internet Computer (ICP) environments (e.g., production and staging) and is not designed to be transferable to other networks.

File: node-ui/src/hooks/useInternetComputer.tsx (hunk 12 above)
Comment (alenmestrov) [reply node-ui/src/hooks/useInternetComputer.tsx]: They are similar calls with the exception of checking for root keys and storing client keys in login() request.

File: node-ui/src/hooks/useInternetComputer.tsx (hunk 10 above)
Comment (alenmestrov) [reply node-ui/src/hooks/useInternetComputer.tsx]: There is already defined return type here (void) on line 134.

File: node-ui/src/api/dataSource/NodeDataSource.ts (hunk 3 above)
Comment (chefsale) [reply node-ui/src/api/dataSource/NodeDataSource.ts]: ICP maybe? 😂

File: crates/storage/README.md
Comment (miraclx) [inline crates/storage/README.md:1]: Do you think we can leave this out? do it in a separate PR maybe? Not saying we must, just checking.

File: crates/server/src/verifywalletsignatures/internetcomputer.rs [hunk 16]
```diff
@@ -0,0 +1,185 @@
...
+/// # Arguments
+/// - `sep`: The domain separator to prepend.
+/// - `bytes`: The message to append.
+///
+/// # Returns
+/// A vector combining the domain separator and the message.
+fn msg_with_domain(sep: &[u8], bytes: &[u8]) -> Vec<u8> {
+    let mut msg = vec![sep.len() as u8];
+    msg.append(&mut sep.to_vec());
+    msg.append(&mut bytes.to_vec());
```
Comment (miraclx) [inline crates/server/src/verifywalletsignatures/internetcomputer.rs]: `[u8]::to_vec` will allocate

try [`Vec::extend_from_slice`](https://doc.rust-lang.org/std/vec/struct.Vec.html#method.extend_from_slice) instead

File: node-ui/src/api/dataSource/NodeDataSource.ts (hunk 3 above)
Comment (alenmestrov) [reply node-ui/src/api/dataSource/NodeDataSource.ts]: We mentioned in previous implementations that we will use full names of protocols, that is why I went with full name.

File: crates/storage/README.md
Comment (alenmestrov) [reply crates/storage/README.md:1]: sure, I reverted back changes done by lint

File: node-ui/src/api/dataSource/NodeDataSource.ts (hunk 4 above)
Comment (MatejVukosav) [reply node-ui/src/api/dataSource/NodeDataSource.ts]: Ok, can u please extract it to const to indicate what it is

File: node-ui/src/hooks/useICP.tsx [hunk 17]
```diff
@@ -0,0 +1,383 @@
...
+ * This interface provides a hook-based approach to managing the user's authentication status, including requesting node data, signing messages for login or root key addition, and handling network changes.
+ * It manages the underlying state for wallet signatures, user readiness, and interactions with the IC Identity service.
+ *
+ * Key functionalities include:
+ * - Requesting node challenge data for signing.
+ * - Logging in or adding a root key by signing a challenge.
+ * - Switching between different Internet Computer networks (production or staging).
+ * - Managing the user's authentication state and error handling throughout the process.
+ */
+interface useICPReturn {
```
Comment (MatejVukosav) [inline node-ui/src/hooks/useICP.tsx]: camel case

File: node-ui/src/pages/ICP.tsx [hunk 18]
```diff
@@ -0,0 +1,46 @@
+import React, { useEffect, useState } from 'react';
+import ContentWrapper from '../components/login/ContentWrapper';
+import { useNavigate } from 'react-router-dom';
+import { ICP } from '../components/icp/ICP';
+import { useICP } from '../hooks/useICP';
+
+interface ICPProps {
```
Comment (MatejVukosav) [inline node-ui/src/pages/ICP.tsx]: camel case. and everywhere else

**Review by MatejVukosav — APPROVED** [review]
Lgtm. 

Fix failing build and namings

**Review by MatejVukosav — APPROVED** [review]
